### PR TITLE
Enable SAML authentication

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This PR enables SAML Authentication in the React Native app

depends on https://github.com/mattermost/platform/pull/5635

It also fixes a bug where the config and license weren't retrieve once more if the user logged out from the server

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [x] All new/modified APIs include changes to the drivers
- [x] Has UI changes
- [x] Includes text changes and localization file updates

### CODE REVIEW

**Comment by enahum** [comment]
@jarredwit can you help out with fixing the tests?

**Comment by enahum** [comment]
@jarredwitt  can you help out with fixing the tests?

**Comment by enahum** [comment]
@jarredwitt no worries, already fixed it

File: android/app/src/main/java/com/mattermost/MainApplication.java [hunk 1]
```diff
@@ -4,6 +4,8 @@
 import android.util.Log;
 
 import com.facebook.react.ReactApplication;
+import com.psykar.cookiemanager.CookieManagerPackage;
```
Comment (hmhealey) [inline android/app/src/main/java/com/mattermost/MainApplication.java]: Which of these are we actually using? I'm amazed the java compiler doesn't freak out at this due to the same class name being imported twice

Comment (enahum) [inline android/app/src/main/java/com/mattermost/MainApplication.java]: Good catch, actually that was the react-native link, let me take a look and see in detail whats going on here. i'll post back

Comment (enahum) [inline android/app/src/main/java/com/mattermost/MainApplication.java]: @hmhealey remove the unnecessary package

**Comment by enahum** [comment]
Im going to test this on Android and let you all know

File: app/navigation/routes.js [hunk 2]
```diff
@@ -167,6 +177,14 @@ export const Routes = {
         navigationProps: {
             title: {id: 'mobile.routes.user_profile', defaultMessage: 'Profile'}
         }
+    },
+    Saml: {
+        key: 'Saml',
+        transition: RouteTransitions.Horizontal,
+        component: Saml,
+        navigationProps: {
+            title: {id: 'mobile.routes.saml', defaultMessage: 'Single SignOn'}
```
Comment (hmhealey) [inline app/navigation/routes.js]: Is there a server setting to change what SAML is called? We might need to figure out how to customize this without recompiling the app

**Comment by enahum** [comment]
@hmhealey this works fine on Android so its ready to finish review

File: app/scenes/select_server/select_server.js [hunk 3]
```diff
@@ -13,21 +13,29 @@ import Button from 'react-native-button';
...
-import {isValidUrl, stripTrailingSlashes} from 'app/utils/url';
 
 export default class SelectServer extends PureComponent {
     static propTypes = {
+        transition: PropTypes.bool.isRequired,
         serverUrl: PropTypes.string.isRequired,
         server: PropTypes.object.isRequired,
-        actions: PropTypes.object.isRequired
+        configRequest: PropTypes.object.isRequired,
+        licenseRequest: PropTypes.object.isRequired,
```
Comment (hmhealey) [inline app/scenes/select_server/select_server.js]: Since we've moved these two calls here, do we still get the config and license properly when we already have the server URL configured and go through the Root scene?

File: app/navigation/routes.js (hunk 2 above)
Comment (enahum) [inline app/navigation/routes.js]: Do you mean something like `OneLogin`, `OKTA` etc? if that's the case yes we do, but dunno if we want to use that here, maybe we should talk about it in the feature standup meeting

File: app/scenes/select_server/select_server.js (hunk 3 above)
Comment (enahum) [inline app/scenes/select_server/select_server.js]: yes we do, it just to show the loading indicator, cause we need to make sure we have the config and the license for the action to select the login option to work properly

File: app/navigation/routes.js (hunk 2 above)
Comment (hmhealey) [inline app/navigation/routes.js]: Yeah, we might want to discuss how it'll work in the long term since right now we just have the login button text like "With SAML" in the config

File: app/scenes/select_server/select_server.js (hunk 3 above)
Comment (hmhealey) [inline app/scenes/select_server/select_server.js]: Okay, cool

File: app/scenes/login_options/login_options.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        if (config.EnableSignInWithEmail === 'true' || config.EnableSignInWithUsername === 'true') {
+            return (
+                <Button
+                    key='email'
+                    onPress={this.props.actions.goToLogin}
+                    containerStyle={[GlobalStyles.signupButton, {backgroundColor: '#2389d7'}]}
+                >
+                    <FormattedText
+                        id='signup.email'
+                        defaultMessage='Email and Password'
```
Comment (thomchop) [inline app/scenes/login_options/login_options.js]: If SignInWithEmail is disabled but SignInWithUsername is enabled, we'd still be displaying "Email and Password" here. But oh well.

**Comment by thomchop** [comment]
@enahum, @hmhealey, @jarredwitt 
When I run `react-native link`, I've observed that it reformats native config files (iOS & android). For example, in `Info.plist` it converts indentation to spaces and adds closing tags to self-closing elements. That's what happened here. It changed every line in `Info.plist`, and yet we didn't make any intentional changes to that file. It would've been difficult to detect those changes if we had made them. It also creates unnecessary conflicts.

If this was the final time that the file would be reformatted, I'd roll with it. But inevitable, xcode will reformat the file when given a chance. I predict that it'll convert the file back to tab indentation next time we bump the version or otherwise update the config. We can deal with this by changing the formatting preferences in `xcshareddata`, but that might be overridden by developers' own prefs in `xcuserdata`. Until we address that, I think we should revert the formatting changes caused by `react-native link`. That's what I've been doing.

**Comment by thomchop** [comment]
@enahum For future reference, I find it helpful to be granular/atomic with our commits (when feasible). There were a lot of changes in the main commit of this PR (`"enable saml authentication"`). I think there were enough discreet changes that this commit could've been broken out into half a dozen or so commits. Doing so is tedious but it really helps when it comes to reviewing/testing.

For example, I would've made a commit just for: `Use full-screen loader in SelectServer`.
I feel especially strong that we should make separate commits for unrelated fixes, even if they're micro.  For example. the change to `app/actions/storage/index.js` wasn't necessary for this PR but it fixed an existing bug which I was almost gonna make a ticket for. If that was its own commit it would've been more likely to register that it was indeed an issue and it was fixed.

cc @jarredwitt, @hmhealey 

How do you guys feel about this? Is this valuable food for thought or am I being excessively methodical about this?

File: assets/base/i18n/en.json [hunk 5]
```diff
@@ -1511,9 +1511,11 @@
   "mobile.loading_channels": "Loading Channels...",
   "mobile.loading_members": "Loading Members...",
   "mobile.request.invalid_response": "Received invalid response from the server.",
+  "mobile.login_options.chooseTitle": "Choose your login method",
```
Comment (thomchop) [inline assets/base/i18n/en.json]: What case convention are we using for our i18n ids? For existing strings in the webapp, we were using snake case. For those that have been added in this repo, we've been using a mix of that and camel case. In this particular one we're using both.
cc @hmhealey @jarredwitt

File: ios/Mattermost.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -582,9 +602,11 @@
 				TargetAttributes = {
 					00E356ED1AD99517003FC87E = {
 						CreatedOnToolsVersion = 6.2;
+						DevelopmentTeam = UQ8HT4Q2XM;
```
Comment (thomchop) [inline ios/Mattermost.xcodeproj/project.pbxproj]: Did we mean to change `DevelopmentTeam` here? I'm sure xcode set it automatically, but I don't think it's necessary. We've been getting by just fine with only declaring `DEVELOPMENT_TEAM` in `Info.plist`. Now there will be a couple more places I have to change if I'm gonna run this on my own device. (I have a stash with my own signing identity declarations that I have to reapply every time I wanna run on an iOS device.)

File: test/setup.js [hunk 7]
```diff
@@ -11,10 +11,16 @@ import path from 'path';
 import register from 'babel-core/register';
 import chai from 'chai';
 import chaiEnzyme from 'chai-enzyme';
+import React from 'react-native-mock';
```
Comment (enahum) [inline test/setup.js]: yup no problem

File: app/actions/views/select_server.js [hunk 8]
```diff
@@ -12,6 +13,24 @@ export function handleServerUrlChanged(serverUrl) {
     };
 }
 
+export function handleLoginOptions() {
+    return async (dispatch, getState) => {
+        const {config, license} = getState().entities.general;
+
+        let options = 0;
+        if (config.EnableSaml === 'true' && license.IsLicensed === 'true' && license.SAML === 'true') {
+            options++;
```
Comment (enahum) [inline app/actions/views/select_server.js]: it would be clearer but then we want to have it the other way for when we add other sign in methods

File: app/navigation/routes.js [hunk 9]
```diff
@@ -21,7 +22,8 @@ import {
     SelectServer,
     SelectTeam,
     Thread,
-    UserProfile
+    UserProfile,
+    Saml
```
Comment (enahum) [inline app/navigation/routes.js]: no problem

File: app/scenes/select_server/select_server.js [hunk 10]
```diff
@@ -72,6 +84,15 @@ export default class SelectServer extends PureComponent {
     };
 
     render() {
+        const {serverUrl, server, configRequest, licenseRequest} = this.props;
+        if (server.status === RequestStatus.STARTED ||
+            configRequest.status === RequestStatus.STARTED ||
+            licenseRequest.status === RequestStatus.STARTED) {
```
Comment (enahum) [inline app/scenes/select_server/select_server.js]: can do

File: app/scenes/saml/saml.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        actions: PropTypes.shape({
+            handleSuccessfulLogin: PropTypes.func.isRequired,
+            setStoreFromLocalData: PropTypes.func.isRequired,
+            goToLoadTeam: PropTypes.func.isRequired
+        }).isRequired
+    };
+
+    onNavigationStateChange(navState) {
+        const {url} = navState;
+        if (url.indexOf('/login/sso/saml') !== -1 && url.indexOf('mobile') === -1) {
```
Comment (enahum) [inline app/scenes/saml/saml.js]: yeap no problem

File: app/actions/views/select_server.js (hunk 8 above)
Comment (hmhealey) [inline app/actions/views/select_server.js]: I think @thomchop's method matches the webapp better. We frequently do stuff like
```javascript
const gitlabEnabled = ...;
const ldapEnabled = ...;
const samlEnabled = ...;

if (gitlabEnabled || ldapEnabled || samlEnabled) {
    ...
}
```

And I also agree with `+=` over `++` outside of loop counters

File: app/scenes/saml/saml.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    render() {
+        return (
+            <WebView
+                source={{uri: `${this.props.serverUrl}/login/sso/saml?action=mobile`}}
+                javaScriptEnabledAndroid={true}
+                automaticallyAdjustContentInsets={false}
+                scalesPageToFit={true}
+                onNavigationStateChange={this.onNavigationStateChange.bind(this)}
```
Comment (hmhealey) [inline app/scenes/saml/saml.js]: Agreed. We should avoid binding in the render function. I was actually considering turning on the ESLint error that causes that to be an error

File: app/scenes/select_server/select_server_container.js [hunk 13]
```diff
@@ -5,16 +5,25 @@ import {bindActionCreators} from 'redux';
...
-import {getPing} from 'service/actions/general';
-import {goToLogin} from 'app/actions/navigation';
+import {getPing, resetPing} from 'service/actions/general';
+import {RequestStatus} from 'service/constants';
 import * as SelectServerActions from 'app/actions/views/select_server';
 
 import SelectServer from './select_server';
 
 function mapStateToProps(state) {
+    const {config: configRequest, license: licenseRequest} = state.requests.general;
```
Comment (hmhealey) [inline app/scenes/select_server/select_server_container.js]: Yeah, that makes a lot of sense. It'll save us a fair bit of typing in the long run, and keep things more consistent

**Comment by enahum** [comment]
I'll make the changes and rebase this soon

File: app/scenes/login_options/login_options.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        const {config, license} = this.props;
+        if (config.EnableSaml === 'true' && license.IsLicensed === 'true' && license.SAML === 'true') {
+            return (
+                <Button
+                    key='saml'
+                    onPress={this.props.actions.goToSaml}
+                    containerStyle={[GlobalStyles.signupButton, {backgroundColor: '#34a28b'}]}
+                >
+                    <Text
+                        style={[GlobalStyles.signupButtonText, {color: 'white'}]}
```
Comment (enahum) [inline app/scenes/login_options/login_options.js]: yes!!

File: app/scenes/select_server/select_server_container.js (hunk 13 above)
Comment (enahum) [inline app/scenes/select_server/select_server_container.js]: that can be another PR for sure

File: assets/base/i18n/en.json (hunk 5 above)
Comment (enahum) [inline assets/base/i18n/en.json]: about keeping this file sorted we don't need to do anything here as the script that is going to merge this into the translation server and then submit a PR back to this repo with the latests translations will take care of that ;)

Comment (enahum) [inline assets/base/i18n/en.json]: we should use snake case, I'm going to change this one
